rfc/2022/2022-11-02_TechTeam4_Preview_Environment_Architecture.md

```diff
@@ -0,0 +1,97 @@
+
+The new architecture, known as Preview Environments, will provide VFS teams with a capability that allows greater customization and the ability to determine for themselves what they need to effectively preview changes and collaborate with their teams.
+
+## Design
+Based on discovery, we determined there were different use cases desired from different teams. This raised the need for allowing VFS developers more control in setting up their ephemeral environment, while keeping the ability to utilize those environments as accessible as possible to facilitate internal testing, stakeholder reviews, and user testing with veterans.
+
+For the MVP of the preview environment architecture, the user will interact with GitHub normally and the preview environment will be created with default settings. As a developer pushes their branch to `vets-website` or `vets-api`, the preview environment architecture will detect the branch and begin the process of deploying the code in the branch to a preview environment. Once the continuous integration workflow completes successfully, a request to deploy the preview environment triggers a GitHub Actions workflow. The GHA workflow writes the dynamically generated preview environment configuration to the `app-manifest` repository. EKS detects the change in the `devops` repository and deploys the newly defined environment.
+
+A diagram is submitted [below](#Diagrams) as additional explanation.
+
+**MVP Feature Set:**
+1. Preview Environment MVP will create ephemeral environments that can be accessed without SOCKS
+2. Preview Environment MVP will operate on the EKS infrastructure and paradigm
+3. Preview Environment MVP will automatically create a Preview Environment when a PR is created
+4. Preview Environment MVP will utilize mock test data
+
+### Post-MVP Design: Additional Features
+For reference, there are numerous additional features planned for inclusion in Post-MVP iterations. Specifically, we will begin adding the abilities to specify custom settings like: custom repository configurations, selection of different types of test data, connectivity to lower environment endpoints for external partners. These features will be vetted and prioritized as development progresses.
+
+## Risks
+* Removing the ability for a preview environment to integrate with VA enterprise lower environments may initially reduce the quality of the experience of preview environments for some users. These risks will lessen over time as additional functionality is added into preview environments and this capability is restored.
+* If there is an overlap between the creation of review instances and the creation of preview environments there could be increased resource usage and cost while both products are supported.
```

Comment:
Has there been any thought on rollout to help ease this? Not essential at this point but more curious what the thought of that looks like.

Reply:
In order to ease the transition to the new architecture, we were planning to run both side by side for a short period of time. In case there are issues with the new PE architecture, users would be able to fall back to review instances. Once we have some data indicating the PE deployments are largely successful then we could decommission the RI system.